vault/logical_system.go

```diff
@@ -2522,6 +2488,57 @@ func (b *SystemBackend) handleEnableAuth(ctx context.Context, req *logical.Reque
 	return resp, nil
 }
 
+func (b *SystemBackend) validateVersion(ctx context.Context, version string, pluginName string, pluginType consts.PluginType) (string, *logical.Response, error) {
+	switch version {
+	case "":
+		var err error
+		version, err = selectPluginVersion(ctx, b.System(), pluginName, pluginType)
+		if err != nil {
+			return "", nil, err
+		}
+
+		if version != "" {
+			b.logger.Debug("pinning plugin version", "plugin type", pluginType.String(), "plugin name", pluginName, "plugin version", version)
+		}
+	default:
+		semanticVersion, err := semver.NewVersion(version)
+		if err != nil {
+			return "", logical.ErrorResponse("version %q is not a valid semantic version: %s", version, err), nil
+		}
+
+		// Canonicalize the version.
+		version = "v" + semanticVersion.String()
+	}
+
+	// if a non-builtin version is requested for a builtin plugin, return an error
+	if version != "" {
+		switch pluginType {
+		case consts.PluginTypeSecrets:
+			aliased, ok := mountAliases[pluginName]
+			if ok {
+				pluginName = aliased
+			}
+			if _, ok = b.Core.logicalBackends[pluginName]; ok {
+				if version != versions.GetBuiltinVersion(pluginType, pluginName) {
```

Comment:
One more thought: In the `else` case, should we then set version back to `""`? If we store e.g. `1.12.0+builtin.vault` in `mountEntry.Version`, and then upgrade Vault to 1.12.1, then the Version field becomes a bit misleading, whereas I think the RunningVersion field will get updated just fine on its own?

Reply:
Oh interesting. Yes, that makes sense. I'll set it to `""` in the else cases here.

Reply:
On second thought, this complicates things, since later we copy the version field to runningversion, so changing the version here to `""` will cause both of them to be blank, which might be confusing?